### PR TITLE
[flang] Generate fir.do_loop reduce from DO CONCURRENT REDUCE clause

### PR DESCRIPTION
**Pull request by khaki3**
Derived from #92480. This PR updates the lowering process of DO CONCURRENT to support F'2023 REDUCE clause. The structure `IncrementLoopInfo` is extended to have both reduction operations and symbols in `reduceSymList`. The function `getConcurrentControl` constructs `reduceSymList` for the innermost loop. Finally, `genFIRIncrementLoopBegin` builds `fir.do_loop` with reduction operands.

### CODE REVIEW

**Comment by khaki3** [comment]
As DO CONCURRENT is not necessarily concurrent, it is sufficient to put a reduce clause only on the innermost loop for propagating reduction semantics. The scope of reduction is analyzable.

File: flang/lib/Lower/Bridge.cpp [hunk 1]
```diff
@@ -1741,6 +1744,36 @@ class FirConverter : public Fortran::lower::AbstractConverter {
...
+      return fir::ReduceOperationEnum::MIN;
+    case Fortran::parser::ReductionOperator::Operator::Iand:
+      return fir::ReduceOperationEnum::IAND;
+    case Fortran::parser::ReductionOperator::Operator::Ior:
+      return fir::ReduceOperationEnum::IOR;
+    case Fortran::parser::ReductionOperator::Operator::Ieor:
+      return fir::ReduceOperationEnum::EIOR;
+    }
+    fir::emitFatalError(toLocation(), "illegal reduction operator");
+    return fir::ReduceOperationEnum::Add;
```
Comment (clementval) [inline flang/lib/Lower/Bridge.cpp:1774]: You could use llvm_unreachable if you want to disable the warning

Comment (khaki3) [reply flang/lib/Lower/Bridge.cpp]: Nice!

**Review by clementval — APPROVED** [review]
LGTM

**Pull request merged**